middle_end/flambda2/from_lambda/closure_conversion.ml

```diff
@@ -1578,6 +1482,244 @@ let close_let_rec acc env ~function_declarations
       named ~body
     |> Expr_with_acc.create_let
 
+let wrap_partial_application acc env apply_continuation (apply : IR.apply)
+    approx args missing_args ~arity ~num_trailing_local_params
+    ~contains_no_escaping_local_allocs =
+  (* In case of partial application, creates a wrapping function from scratch to
+     allow inlining and lifting *)
+  let wrapper_id = Ident.create_local ("partial_" ^ Ident.name apply.func) in
+  let closure_id =
+    Closure_id.wrap
+      (Compilation_unit.get_current_exn ())
+      (Variable.create_with_same_name_as_ident wrapper_id)
+  in
+  let args_arity = List.length args in
+  let params =
+    List.mapi
+      (fun n _ ->
+        ( Ident.create_local ("param" ^ string_of_int (args_arity + n)),
+          Lambda.Pgenval ))
+      missing_args
+  in
+  let return_continuation = Continuation.create ~sort:Return () in
+  let exn_continuation =
+    IR.{ exn_handler = Continuation.create (); extra_args = [] }
+  in
+  let all_args = args @ List.map (fun (a, _) -> IR.Var a) params in
+  let fbody acc env =
+    close_exact_or_unknown_apply acc env
+      { apply with
+        kind = Function;
+        args = all_args;
+        continuation = return_continuation;
+        exn_continuation
+      }
+      (Some approx)
+  in
+  let attr =
+    Lambda.
+      { inline = Default_inline;
+        specialise = Default_specialise;
+        local = Default_local;
+        is_a_functor = false;
+        stub = false
+      }
+  in
+  let free_idents_of_body =
+    List.fold_left
+      (fun ids -> function IR.Var id -> Ident.Set.add id ids | _ -> ids)
+      (Ident.Set.singleton apply.func)
+      all_args
+  in
+  let closure_alloc_mode, num_trailing_local_params =
+    let num_leading_heap_params =
+      List.length arity - num_trailing_local_params
+    in
+    if args_arity <= num_leading_heap_params
+    then Lambda.Alloc_heap, num_trailing_local_params
+    else
+      let num_supplied_local_args = args_arity - num_leading_heap_params in
+      Lambda.Alloc_local, num_trailing_local_params - num_supplied_local_args
+  in
+  let function_declarations =
+    [ Function_decl.create ~let_rec_ident:(Some wrapper_id) ~closure_id
+        ~kind:(Lambda.Curried { nlocal = num_trailing_local_params })
+        ~params ~return:Lambda.Pgenval ~return_continuation ~exn_continuation
```

Comment:
Same comment as above about the return arity: we could recover it from the code's return arity, but we need a conversion.